src/stream.ts

```diff
@@ -107,203 +99,123 @@ export class YamuxStream implements Stream {
-        this.sendWindowUpdate()
-        yield val
-      }
-    } catch (err) {
-      const errCode = (err as { code: string }).code
-      if (errCode !== ERR_STREAM_ABORT) {
-        this.log?.error('stream source error id=%s', this._id, err)
-        throw err
-      }
-    }
-  }
-
-  close (): void {
-    this.log?.trace('stream close id=%s', this._id)
-    this.closeRead()
-    this.closeWrite()
-  }
-
-  closeRead (): void {
-    if (this.state === StreamState.Finished) {
-      return
-    }
-
-    if (this.readState !== HalfStreamState.Open) {
-      return
-    }
-
-    this.log?.trace('stream close read id=%s', this._id)
-
-    this.readState = HalfStreamState.Closed
-
-    // close the source
-    this.sourceInput.end()
+  /**
+   * Send a message to the remote muxer informing them a new stream is being
+   * opened
+   */
+  async sendNewStream (): Promise<void> {
 
-    // If the both read and write are closed, finish it
-    if (this.writeState !== HalfStreamState.Open) {
-      this.finish()
-    }
   }
 
-  closeWrite (): void {
-    if (this.state === StreamState.Finished) {
-      return
-    }
-
-    if (this.writeState !== HalfStreamState.Open) {
-      return
-    }
+  /**
+   * Send a data message to the remote muxer
+   */
+  async sendData (buf: Uint8ArrayList, options: AbortOptions = {}): Promise<void> {
+    buf = buf.sublist()
+
+    // send in chunks, waiting for window updates
+    while (buf.byteLength !== 0) {
+      // wait for the send window to refill
+      if (this.sendWindowCapacity === 0) {
+        await this.waitForSendWindowCapacity(options)
+      }
 
-    this.log?.trace('stream close write id=%s', this._id)
+      // check we didn't close while waiting for send window capacity
+      if (this.status !== 'open') {
+        return
+      }
 
-    this.writeState = HalfStreamState.Closed
+      // send as much as we can
+      const toSend = Math.min(this.sendWindowCapacity, this.config.maxMessageSize - HEADER_LENGTH, buf.length)
+      const flags = this.getSendFlags()
 
-    this.sendClose()
+      this.sendFrame({
+        type: FrameType.Data,
+        flag: flags,
+        streamID: this._id,
+        length: toSend
+      }, buf.subarray(0, toSend))
 
-    // close the sink
-    this.abortController.abort()
+      this.sendWindowCapacity -= toSend
 
-    // If the both read and write are closed, finish it
-    if (this.readState !== HalfStreamState.Open) {
-      this.finish()
+      buf.consume(toSend)
     }
   }
 
-  abort (err?: Error): void {
-    switch (this.state) {
-      case StreamState.Finished:
-        return
-      case StreamState.Init:
-        // we haven't sent anything, so we don't need to send a reset.
-        break
-      case StreamState.SYNSent:
-      case StreamState.SYNReceived:
-      case StreamState.Established:
-        // at least one direction is open, we need to send a reset.
-        this.sendReset()
-        break
```

Comment:
Is this behavior still respected?

Reply:
The stream is opened on the remote by sending the first window update.  This happens in the `.newStream` method on the yamux muxer before the stream is returned to the called, so there's no way to call `.abort(err)` on the stream before the state has gone from `StreamState.Init` to `StreamState.SYNSent`.

Similarly `.state` is updated to `StreamState.Finished` after the readable and writable ends are both closed at which point `.abort(err)` is a no-op so this method is not called.